--- a/impl/aspaper-server/build.gradle.kts
+++ b/impl/aspaper-server/build.gradle.kts
@@ -32,7 +_,22 @@
         }
     }
 
-    activeFork = fork
+    val kitpvpSlime = forks.register("kitpvpslime") {
+        forks = fork
+        upstream.patchRepo("paperServer") {
+            upstreamRepo = fork.patchedRepo("paperServer")
+            patchesDir = rootDirectory.dir("kitpvpslime-server/paper-patches")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+        upstream.patchDir("kitpvpSlimeServer") {
+            upstreamPath = "impl/aspaper-server"
+            excludes = setOf("src/minecraft", "paper-patches", "minecraft-patches", "build.gradle.kts", "build.gradle.kts.patch")
+            patchesDir = rootDirectory.dir("kitpvpslime-server/aspaper-patches")
+            outputDir = rootDirectory.dir("aspaper-server")
+        }
+    }
+
+    activeFork = kitpvpSlime
 
     spigot {
         buildDataRef = "3edaf46ec1eed4115ce1b18d2846cded42577e42"
@@ -118,10 +_,14 @@
     main {
         java { srcDir("../paper-server/src/main/java") }
         resources { srcDir("../paper-server/src/main/resources") }
+        java { srcDir("../aspaper-server/src/main/java") }
+        resources { srcDir("../aspaper-server/src/main/resources") }
     }
     test {
         java { srcDir("../paper-server/src/test/java") }
         resources { srcDir("../paper-server/src/test/resources") }
+        java { srcDir("../aspaper-server/src/test/java") }
+        resources { srcDir("../aspaper-server/src/test/resources") }
     }
 }
 
@@ -145,7 +_,7 @@
 }
 
 dependencies {
-    implementation(project(":impl:aspaper-api")) //ASP
+    implementation(project(":kitpvpslime-api")) //ASP
     implementation(project(":core")) //ASP
     implementation("commons-io:commons-io:2.11.0")
     implementation("ca.spottedleaf:concurrentutil:0.0.3")
