From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NotStevy <notstevy@ultrabuildmc.de>
Date: Mon, 19 Aug 2024 11:28:26 +0200
Subject: [PATCH] Build changes


diff --git a/build.gradle.kts b/build.gradle.kts
index 3a1cad6046387bacf0e7ef83be8ee2e1a92137b7..c736b98a12edd3ee58f8f519e3c0f7a1a43fa2b9 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -3,6 +3,6 @@ plugins {
 
 dependencies {
     compileOnly(project(":api"))
-    compileOnly(project(":slimeworldmanager-api"))
+    compileOnly(project(":kitpvpslime-api")) // kitpvp rename
     implementation("com.github.luben:zstd-jni:1.5.2-2")
 }
