From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: notstevy <notstevy@ultrabuildmc.de>
Date: Sun, 6 Apr 2025 11:43:32 +0200
Subject: [PATCH] Display Player - Porting changes from old fork


diff --git a/src/main/java/org/bukkit/entity/EntityType.java b/src/main/java/org/bukkit/entity/EntityType.java
index 29d22511660060c182314e23ad849bca34611717..195921741e5c55fae783bf354659148bba08fce3 100644
--- a/src/main/java/org/bukkit/entity/EntityType.java
+++ b/src/main/java/org/bukkit/entity/EntityType.java
@@ -338,10 +338,13 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
      */
     LIGHTNING_BOLT("lightning_bolt", LightningStrike.class, -1),
     PLAYER("player", Player.class, -1, false),
+    // kitpvp start - Display Player
+    DISPLAY_PLAYER(NamespacedKey.fromString("kitpvp:display_player"), world.kitpvp.slime.api.entity.DisplayPlayer.class, -1, true),
     /**
      * An unknown entity without an Entity Class
      */
-    UNKNOWN(null, null, -1, false);
+    UNKNOWN((String) null, null, -1, false);
+    // kitpvp end - DisplayPlayer
 
     private final String name;
     private final Class<? extends Entity> clazz;
@@ -376,6 +379,17 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
         this.key = (name == null) ? null : NamespacedKey.minecraft(name);
     }
 
+    // kitpvp start - DisplayPlayer
+    private EntityType(@NotNull NamespacedKey key, /*@Nullable*/ Class<? extends Entity> clazz, int typeId, boolean independent) {
+        this.name = key.asMinimalString();
+        this.clazz = clazz;
+        this.typeId = (short) typeId;
+        this.independent = independent;
+        this.living = clazz != null && LivingEntity.class.isAssignableFrom(clazz);
+        this.key = key;
+    }
+    // kitpvp end - DisplayPlayer
+
     /**
      * Gets the entity type name.
      *
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 494dca2ee48a03953d47050b178496df12bc48c5..78edb96332b1c2346ece50f6f453c1dd0d748e59 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -62,7 +62,7 @@ import org.jspecify.annotations.Nullable;
  * Represents a player, connected or not
  */
 @NullMarked
-public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient { // Paper
+public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient, world.kitpvp.slime.api.entity.DisplayPlayer { // Paper  // kitpvp - extends DisplayPlayer
 
     // Paper start
     @Override
