From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: notstevy <notstevy@ultrabuildmc.de>
Date: Thu, 17 Apr 2025 12:22:11 +0200
Subject: [PATCH] Respawn Screen API


diff --git a/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java b/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java
index 6316a2f1d2dc0314397e33e6dbd354fb8bc50541..1f11866b36aa060b71db6e788cfb61dd534a00d4 100644
--- a/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java
+++ b/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java
@@ -23,6 +23,7 @@ public class PlayerDeathEvent extends EntityDeathEvent {
     private Component deathMessage;
     private Component deathScreenMessageOverride = null;
     private boolean doExpDrop;
+    private boolean cancelRespawnScreen = false; // kitpvp - Respawn Screen API
     private boolean keepLevel = false;
     private boolean keepInventory = false;
     @Deprecated
@@ -341,4 +342,22 @@ public class PlayerDeathEvent extends EntityDeathEvent {
     public List<ItemStack> getItemsToKeep() {
         return this.itemsToKeep;
     }
+
+    // kitpvp start - Respawn Screen API
+    /**
+     * Gets if the respawn screen should be displayed to the Player.
+     * @return True if the respawn screen should not be shown
+     */
+    public boolean shouldCancelRespawnScreen() {
+        return cancelRespawnScreen;
+    }
+    /**
+     * Sets if the respawn screen should be displayed to the Player.
+     * <p>
+     * <b>This only works if the respawn world and the original world are the same (may break if otherwise).</b>
+     */
+    public void cancelRespawnScreen() {
+        this.cancelRespawnScreen = true;
+    }
+    // kitpvp end - Respawn Screen API
 }
