From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: notstevy <notstevy@ultrabuildmc.de>
Date: Fri, 26 Sep 2025 16:42:44 +0200
Subject: [PATCH] Predicatable Enchantments Modifying enchantments (such as
 Unbreaking) to not use random, making pvp more deterministic

Signed-off-by: notstevy <notstevy@ultrabuildmc.de>

diff --git a/src/main/java/org/bukkit/GameRule.java b/src/main/java/org/bukkit/GameRule.java
index ff5893a7144bc3d740675f6833424c858598b9b4..4d86c25a469606ebe3d51e8b4c97875f388c6530 100644
--- a/src/main/java/org/bukkit/GameRule.java
+++ b/src/main/java/org/bukkit/GameRule.java
@@ -310,6 +310,13 @@ public final class GameRule<T> implements net.kyori.adventure.translation.Transl
      */
     public static final GameRule<Boolean> LOCATOR_BAR = new GameRule<>("locatorBar", Boolean.class);
 
+    // kitpvp start - Constant Enchantments
+    /**
+     * KitPvP: Configures whether enchantments like unbreaking should be random or deterministic
+     */
+    public static final GameRule<Boolean> CONSTANT_ENCHANTMENTS = new GameRule<>("constantEnchantments", Boolean.class);
+    // kitpvp end - Constant Enchantments
+
     // All GameRules instantiated above this for organizational purposes
     private final String name;
     private final Class<T> type;
