--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -664,7 +_,11 @@
             null, // Paper - rewrite chunk system
             () -> server.overworld().getDataStorage()
         );
-        this.chunkSource.getGeneratorState().ensureStructuresGenerated();
+        // kitpvp start - optimize world loading
+        if(!(this instanceof com.infernalsuite.asp.level.SlimeLevelInstance)) {
+            this.chunkSource.getGeneratorState().ensureStructuresGenerated();
+        }
+        // kitpvp end - optimize world loading
         this.portalForcer = new PortalForcer(this);
         this.updateSkyBrightness();
         this.prepareWeather();
