--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -204,16 +_,18 @@
         this.mainThread = Thread.currentThread();
         Path path = levelStorageAccess.getDimensionPath(level.dimension()).resolve("data");
 
-        try {
-            FileUtil.createDirectoriesSafe(path);
-        } catch (IOException var15) {
-            LOGGER.error("Failed to create dimension data storage directory", (Throwable)var15);
-        }
-
+        // kitpvp start - optimize world loading
         //ASP start - No dimension data storage
         if(level instanceof com.infernalsuite.asp.level.SlimeLevelInstance) {
             this.dataStorage = new com.infernalsuite.asp.level.ReadOnlyDimensionDataStorage(new SavedData.Context(level), path, fixerUpper, level.registryAccess());
         } else {
+            try {
+                FileUtil.createDirectoriesSafe(path);
+            } catch (IOException var15) {
+                LOGGER.error("Failed to create dimension data storage directory", (Throwable)var15);
+            }
+            // kitpvp end - optimize world loading
+
             this.dataStorage = new DimensionDataStorage(new SavedData.Context(level), path, fixerUpper, level.registryAccess());
         }
         //ASP end - No dimension data storage
