--- a/net/minecraft/world/entity/EntitySelector.java
+++ b/net/minecraft/world/entity/EntitySelector.java
@@ -14,7 +_,14 @@
     public static final Predicate<Entity> NO_CREATIVE_OR_SPECTATOR = entity -> !(
         entity instanceof Player player && (entity.isSpectator() || player.isCreative())
     );
-    public static final Predicate<Entity> NO_SPECTATORS = entity -> !entity.isSpectator();
+    // // kitpvp start - Extend collides API
+    public static final Predicate<Entity> NO_SPECTATORS = entity -> {
+        if (entity instanceof LivingEntity livingEntity && !livingEntity.collides)
+            return false;
+
+        return !entity.isSpectator();
+    };
+    // kitpvp end
     public static final Predicate<Entity> CAN_BE_COLLIDED_WITH = NO_SPECTATORS.and(entity -> entity.canBeCollidedWith(null));
     public static final Predicate<Entity> CAN_BE_PICKED = NO_SPECTATORS.and(Entity::isPickable);
     // Paper start - Ability to control player's insomnia and phantoms
