--- a/net/minecraft/world/entity/EntitySelector.java
+++ b/net/minecraft/world/entity/EntitySelector.java
@@ -13,7 +_,14 @@
     public static final Predicate<Entity> CONTAINER_ENTITY_SELECTOR = entity -> entity instanceof Container && entity.isAlive();
     public static final Predicate<Entity> NO_CREATIVE_OR_SPECTATOR = entity -> !(entity instanceof Player)
         || !entity.isSpectator() && !((Player)entity).isCreative();
-    public static final Predicate<Entity> NO_SPECTATORS = entity -> !entity.isSpectator();
+    // kitpvp start - Extend collides API
+    public static final Predicate<Entity> NO_SPECTATORS = entity -> {
+        if(entity instanceof LivingEntity livingEntity && !livingEntity.collides)
+            return false;
+
+        return !entity.isSpectator();
+    };
+    // kitpvp end
     public static final Predicate<Entity> CAN_BE_COLLIDED_WITH = NO_SPECTATORS.and(Entity::canBeCollidedWith);
     public static final Predicate<Entity> CAN_BE_PICKED = NO_SPECTATORS.and(Entity::isPickable);
     // Paper start - Ability to control player's insomnia and phantoms
