From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: notstevy <notstevy@ultrabuildmc.de>
Date: Sun, 6 Apr 2025 11:43:31 +0200
Subject: [PATCH] Display Player


diff --git a/net/minecraft/network/syncher/SynchedEntityData.java b/net/minecraft/network/syncher/SynchedEntityData.java
index 3d90f9f1ac1bd281edf6bb0f93ea821657d5bd2f..4d366caf1b6b18fd1911f4e16c7daa052bf58e12 100644
--- a/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/net/minecraft/network/syncher/SynchedEntityData.java
@@ -37,6 +37,19 @@ public class SynchedEntityData {
             }
         }
 
+        // kitpvp start
+        if(clazz == world.kitpvp.slime.entity.DisplayPlayer.class)
+            throw new RuntimeException("Cannot register a SynchedEntityData for the 'world.kitpvp.slime.entity.DisplayPlayer'! It must be synchronous with the 'net.minecraft.world.entity.player.Player'");
+        if(clazz == net.minecraft.world.entity.player.Player.class) {
+            var displayPlayerLastId = ID_REGISTRY.getLastIdFor(world.kitpvp.slime.entity.DisplayPlayer.class);
+            var playerLastId = ID_REGISTRY.getLastIdFor(net.minecraft.world.entity.player.Player.class);
+            if(displayPlayerLastId != playerLastId)
+                throw new RuntimeException("'world.kitpvp.slime.entity.DisplayPlayer' ("+ displayPlayerLastId + ") is not in sync with the 'net.minecraft.world.entity.player.Player' (" + playerLastId + ")");
+
+            ID_REGISTRY.define(world.kitpvp.slime.entity.DisplayPlayer.class);
+        }
+        // kitpvp end
+
         int i = ID_REGISTRY.define(clazz);
         if (i > 254) {
             throw new IllegalArgumentException("Data value id is too big with " + i + "! (Max is 254)");
diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 0fb253aa55a24b56b17f524b3261c5b75c7d7e59..c51f97ad33b85a839b857d487091354f8e8fa663 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -322,6 +322,17 @@ public class ServerEntity {
             // CraftBukkit end
         }
 
+        // kitpvp start - DisplayPlayer
+        if(this.entity instanceof world.kitpvp.slime.entity.DisplayPlayer displayPlayer) {
+            consumer.accept(
+                new net.minecraft.network.protocol.game.ClientboundPlayerInfoUpdatePacket(
+                    java.util.EnumSet.of(net.minecraft.network.protocol.game.ClientboundPlayerInfoUpdatePacket.Action.ADD_PLAYER),
+                    displayPlayer.getInfoUpdateEntry(player)
+                )
+            );
+        }
+        // kitpvp end - DisplayPlayer
+
         Packet<ClientGamePacketListener> addEntityPacket = this.entity.getAddEntityPacket(this);
         consumer.accept(addEntityPacket);
         if (this.trackedDataValues != null) {
diff --git a/net/minecraft/world/entity/EntityType.java b/net/minecraft/world/entity/EntityType.java
index 49201d6664656ebe34c84c1c84b5ea4878729062..cadce0075dd250b350f12d3043eb1777212460c9 100644
--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1043,6 +1043,18 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
             .clientTrackingRange(32)
             .updateInterval(2)
     );
+    // kitpvp start - DisplayPlayer
+    public static final EntityType<world.kitpvp.slime.entity.DisplayPlayer> DISPLAY_PLAYER = register(
+        ResourceKey.create(Registries.ENTITY_TYPE, ResourceLocation.fromNamespaceAndPath("kitpvp", "display_player")),
+        EntityType.Builder.<world.kitpvp.slime.entity.DisplayPlayer>of(world.kitpvp.slime.entity.DisplayPlayer::new, MobCategory.MISC)
+            .noSave()
+            .sized(0.6F, 1.8F)
+            .eyeHeight(1.62F)
+            .vehicleAttachment(Player.DEFAULT_VEHICLE_ATTACHMENT)
+            .clientTrackingRange(32)
+            .updateInterval(2)
+    );
+    // kitpvp end - DisplayPlayer
     public static final EntityType<FishingHook> FISHING_BOBBER = register(
         "fishing_bobber",
         EntityType.Builder.<FishingHook>of(FishingHook::new, MobCategory.MISC)
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 735b9ee502538200dd0d25f65f3a0b8acaca4fa9..0f237b9d6e092dabb3dab6de4616c66ad8d3c34c 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -2635,6 +2635,13 @@ public abstract class LivingEntity extends Entity implements Attackable {
         return this.attributes;
     }
 
+    // kitpvp start - DisplayPlayer
+    public void setAttributes(AttributeMap attributeMap) {
+        this.attributes = attributeMap;
+        this.craftAttributes = new CraftAttributeMap(this.attributes);
+    }
+    // kitpvp end - DisplayPlayer
+
     public ItemStack getMainHandItem() {
         return this.getItemBySlot(EquipmentSlot.MAINHAND);
     }
diff --git a/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java b/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
index 33527a1825119f3667fb3c7ccec318f2c7328ec9..80e4a9e81014d7effc6b6d0865af144c7086fa86 100644
--- a/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
+++ b/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
@@ -137,6 +137,7 @@ public class DefaultAttributes {
         .put(EntityType.PIGLIN_BRUTE, PiglinBrute.createAttributes().build())
         .put(EntityType.PILLAGER, Pillager.createAttributes().build())
         .put(EntityType.PLAYER, Player.createAttributes().build())
+        .put(EntityType.DISPLAY_PLAYER, world.kitpvp.slime.entity.DisplayPlayer.createAttributes().build()) // kitpvp
         .put(EntityType.POLAR_BEAR, PolarBear.createAttributes().build())
         .put(EntityType.PUFFERFISH, AbstractFish.createAttributes().build())
         .put(EntityType.RABBIT, Rabbit.createAttributes().build())
