From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: notstevy <notstevy@ultrabuildmc.de>
Date: Thu, 17 Jul 2025 15:49:24 +0200
Subject: [PATCH] Virtual Inventory API

Signed-off-by: notstevy <notstevy@ultrabuildmc.de>

diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 8891baab314cd1a1096c3930e7357b0f368efe1d..26e87fd74cee0f1bb9f26899b06303e56cacfff3 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -302,7 +302,8 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
         @Override
         public void sendSlotChange(AbstractContainerMenu container, int slot, ItemStack itemStack) {
-            ServerPlayer.this.connection.send(new ClientboundContainerSetSlotPacket(container.containerId, container.incrementStateId(), slot, itemStack));
+            if(container != ServerPlayer.this.inventoryMenu || !ServerPlayer.this.hasVirtualInventory()) // kitpvp - Virtual Inventory
+                ServerPlayer.this.connection.send(new ClientboundContainerSetSlotPacket(container.containerId, container.incrementStateId(), slot, itemStack));
         }
 
         @Override
@@ -1900,6 +1901,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.containerMenu.removed(this);
         this.inventoryMenu.transferState(this.containerMenu);
         this.containerMenu = this.inventoryMenu;
+        this.virtualInventory(null); // kitpvp - Virtual Inventory
     }
 
     @Override
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 5bba860a5f913d3a83c1d469f0f4854b6e063a49..36a52f3e21b7c6bb675e33f03d63d982c95af920 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -3064,7 +3064,7 @@ public class ServerGamePacketListenerImpl
                                 click = (packet.buttonNum() == Inventory.SLOT_OFFHAND) ? ClickType.SWAP_OFFHAND : ClickType.NUMBER_KEY;
                                 Slot clickedSlot = this.player.containerMenu.getSlot(slotNum);
                                 if (clickedSlot.mayPickup(this.player)) {
-                                    ItemStack hotbar = this.player.getInventory().getItem(packet.buttonNum());
+                                    ItemStack hotbar = this.player.virtualOrInventory().getItem(packet.buttonNum()); // kitpvp - Virtual Inventory
                                     if ((!hotbar.isEmpty() && clickedSlot.mayPlace(hotbar)) || (hotbar.isEmpty() && clickedSlot.hasItem())) { // Paper - modernify this logic (no such thing as a "hotbar move and readd"
                                         action = InventoryAction.HOTBAR_SWAP;
                                     } else {
@@ -3201,6 +3201,15 @@ public class ServerGamePacketListenerImpl
                             }
                         }
                         // Paper end - cartography item event
+                        // kitpvp start - Virtual Inventory
+                        if (this.player.hasVirtualInventory()) {
+                            if (click == ClickType.SWAP_OFFHAND
+                                || action == InventoryAction.DROP_ALL_SLOT || action == InventoryAction.DROP_ALL_CURSOR
+                                || action == InventoryAction.DROP_ONE_SLOT || action == InventoryAction.DROP_ONE_CURSOR) {
+                                cancelled = true;
+                            }
+                        }
+                        // kitpvp end - Virtual Inventory
 
                         event.setCancelled(cancelled);
                         net.minecraft.world.inventory.AbstractContainerMenu oldContainer = this.player.containerMenu; // SPIGOT-1224
diff --git a/net/minecraft/world/entity/player/Inventory.java b/net/minecraft/world/entity/player/Inventory.java
index a6bb436dc80daf6901dc027a6011ead4b3ed27e2..3d55ea8d32b3bff2baed3d6c35c69059ce9513e1 100644
--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -409,6 +409,7 @@ public class Inventory implements Container, Nameable {
     }
 
     public void placeItemBackInInventory(ItemStack stack, boolean sendPacket) {
+        if(!this.player.isInventoryAllowedToUpdate(this)) sendPacket = false;  // kitpvp - Virtual Inventory
         while (!stack.isEmpty()) {
             int slotWithRemainingSpace = this.getSlotWithRemainingSpace(stack);
             if (slotWithRemainingSpace == -1) {
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 34b8d020140d9ed563b7cbe0a604b9ce2ffe4d49..e2e2ceb5370ec179272d11de8d0e5b1719ef7558 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -181,6 +181,7 @@ public abstract class Player extends LivingEntity {
     protected PlayerEnderChestContainer enderChestInventory = new PlayerEnderChestContainer(this); // CraftBukkit - add "this" to constructor
     public final InventoryMenu inventoryMenu;
     public AbstractContainerMenu containerMenu;
+    private @Nullable Inventory virtualInventory; // kitpvp - Virtual Inventory
     protected FoodData foodData = new FoodData();
     protected int jumpTriggerTime;
     private boolean clientLoaded = false;
@@ -1460,6 +1461,28 @@ public abstract class Player extends LivingEntity {
         return this.inventory;
     }
 
+    // kitpvp start - Virtual Inventory
+    public void virtualInventory(@Nullable Inventory virtualInventory) {
+        var hadVirtualInventory = this.hasVirtualInventory();
+        this.virtualInventory = virtualInventory;
+        if(hadVirtualInventory && virtualInventory == null) {
+            inventoryMenu.sendAllDataToRemote();
+        }
+    }
+
+    public Inventory virtualOrInventory() {
+        return this.virtualInventory != null ? this.virtualInventory : this.getInventory();
+    }
+
+    public boolean isInventoryAllowedToUpdate(Inventory inventory) {
+        return hasVirtualInventory() ? inventory == virtualInventory : inventory == this.inventory;
+    }
+
+    public boolean hasVirtualInventory() {
+        return this.virtualInventory != null;
+    }
+    // kitpvp end - Virtual Inventory
+
     public Abilities getAbilities() {
         return this.abilities;
     }
diff --git a/net/minecraft/world/inventory/AbstractContainerMenu.java b/net/minecraft/world/inventory/AbstractContainerMenu.java
index 96da9b8e3556e58365dc6742f44d6dfd608e2953..522c23bfdbb79dc74e3b61274abf752d0470ecb1 100644
--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -399,7 +399,7 @@ public abstract class AbstractContainerMenu {
     }
 
     private void doClick(int slotId, int button, ClickType clickType, Player player) {
-        Inventory inventory = player.getInventory();
+        Inventory inventory = player.virtualOrInventory(); // kitpvp - Virtual Inventory
         if (clickType == ClickType.QUICK_CRAFT) {
             int i = this.quickcraftStatus;
             this.quickcraftStatus = getQuickcraftHeader(button);
