From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: notstevy <notstevy@ultrabuildmc.de>
Date: Thu, 17 Jul 2025 15:49:24 +0200
Subject: [PATCH] Virtual Inventory API

Signed-off-by: notstevy <notstevy@ultrabuildmc.de>

diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index fd83f0c2715b019e599e3b53707e3421dc4a193f..f73cccc8938ceb4a0d1c7a25308b0e9d78d6696b 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -302,7 +302,8 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
         @Override
         public void sendSlotChange(AbstractContainerMenu container, int slot, ItemStack itemStack) {
-            ServerPlayer.this.connection.send(new ClientboundContainerSetSlotPacket(container.containerId, container.incrementStateId(), slot, itemStack));
+            if(container != ServerPlayer.this.inventoryMenu || !ServerPlayer.this.hasVirtualInventory()) // kitpvp - Virtual Inventory
+                ServerPlayer.this.connection.send(new ClientboundContainerSetSlotPacket(container.containerId, container.incrementStateId(), slot, itemStack));
         }
 
         @Override
@@ -1904,6 +1905,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.containerMenu.removed(this);
         this.inventoryMenu.transferState(this.containerMenu);
         this.containerMenu = this.inventoryMenu;
+        this.virtualInventory(null); // kitpvp - Virtual Inventory
     }
 
     @Override
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index ba29032e84930a753c687e95222e64a7fe728b0d..a736c83e7bdf40d74da243c8da4367aa6d453a3a 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -3066,7 +3066,7 @@ public class ServerGamePacketListenerImpl
                                 click = (packet.buttonNum() == Inventory.SLOT_OFFHAND) ? ClickType.SWAP_OFFHAND : ClickType.NUMBER_KEY;
                                 Slot clickedSlot = this.player.containerMenu.getSlot(slotNum);
                                 if (clickedSlot.mayPickup(this.player)) {
-                                    ItemStack hotbar = this.player.getInventory().getItem(packet.buttonNum());
+                                    ItemStack hotbar = this.player.virtualOrInventory().getItem(packet.buttonNum()); // kitpvp - Virtual Inventory
                                     if ((!hotbar.isEmpty() && clickedSlot.mayPlace(hotbar)) || (hotbar.isEmpty() && clickedSlot.hasItem())) { // Paper - modernify this logic (no such thing as a "hotbar move and readd"
                                         action = InventoryAction.HOTBAR_SWAP;
                                     } else {
@@ -3203,6 +3203,15 @@ public class ServerGamePacketListenerImpl
                             }
                         }
                         // Paper end - cartography item event
+                        // kitpvp start - Virtual Inventory
+                        if (this.player.hasVirtualInventory()) {
+                            if (click == ClickType.SWAP_OFFHAND
+                                || action == InventoryAction.DROP_ALL_SLOT || action == InventoryAction.DROP_ALL_CURSOR
+                                || action == InventoryAction.DROP_ONE_SLOT || action == InventoryAction.DROP_ONE_CURSOR) {
+                                cancelled = true;
+                            }
+                        }
+                        // kitpvp end - Virtual Inventory
 
                         event.setCancelled(cancelled);
                         net.minecraft.world.inventory.AbstractContainerMenu oldContainer = this.player.containerMenu; // SPIGOT-1224
diff --git a/net/minecraft/world/entity/player/Inventory.java b/net/minecraft/world/entity/player/Inventory.java
index a6bb436dc80daf6901dc027a6011ead4b3ed27e2..3d55ea8d32b3bff2baed3d6c35c69059ce9513e1 100644
--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -409,6 +409,7 @@ public class Inventory implements Container, Nameable {
     }
 
     public void placeItemBackInInventory(ItemStack stack, boolean sendPacket) {
+        if(!this.player.isInventoryAllowedToUpdate(this)) sendPacket = false;  // kitpvp - Virtual Inventory
         while (!stack.isEmpty()) {
             int slotWithRemainingSpace = this.getSlotWithRemainingSpace(stack);
             if (slotWithRemainingSpace == -1) {
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 94edb1f01848bffd4e75d7976e2eaf37c207c87c..685905da0ebaabbaa8f3242860e4c6625a66170e 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -181,6 +181,7 @@ public abstract class Player extends LivingEntity {
     protected PlayerEnderChestContainer enderChestInventory = new PlayerEnderChestContainer(this); // CraftBukkit - add "this" to constructor
     public final InventoryMenu inventoryMenu;
     public AbstractContainerMenu containerMenu;
+    private @Nullable Inventory virtualInventory; // kitpvp - Virtual Inventory
     protected FoodData foodData = new FoodData();
     protected int jumpTriggerTime;
     private boolean clientLoaded = false;
@@ -1461,6 +1462,28 @@ public abstract class Player extends LivingEntity {
         return this.inventory;
     }
 
+    // kitpvp start - Virtual Inventory
+    public void virtualInventory(@Nullable Inventory virtualInventory) {
+        var hadVirtualInventory = this.hasVirtualInventory();
+        this.virtualInventory = virtualInventory;
+        if(hadVirtualInventory && virtualInventory == null) {
+            inventoryMenu.sendAllDataToRemote();
+        }
+    }
+
+    public Inventory virtualOrInventory() {
+        return this.virtualInventory != null ? this.virtualInventory : this.getInventory();
+    }
+
+    public boolean isInventoryAllowedToUpdate(Inventory inventory) {
+        return hasVirtualInventory() ? inventory == virtualInventory : inventory == this.inventory;
+    }
+
+    public boolean hasVirtualInventory() {
+        return this.virtualInventory != null;
+    }
+    // kitpvp end - Virtual Inventory
+
     public Abilities getAbilities() {
         return this.abilities;
     }
diff --git a/net/minecraft/world/inventory/AbstractContainerMenu.java b/net/minecraft/world/inventory/AbstractContainerMenu.java
index 06846950348954328c07f64cd9b3359e79a1a468..73d04797a23514f5fb2965213834b0ee60ca9f3c 100644
--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -399,7 +399,7 @@ public abstract class AbstractContainerMenu {
     }
 
     private void doClick(int slotId, int button, ClickType clickType, Player player) {
-        Inventory inventory = player.getInventory();
+        Inventory inventory = player.virtualOrInventory(); // kitpvp - Virtual Inventory
         if (clickType == ClickType.QUICK_CRAFT) {
             int i = this.quickcraftStatus;
             this.quickcraftStatus = getQuickcraftHeader(button);
