From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: notstevy <notstevy@ultrabuildmc.de>
Date: Sun, 6 Apr 2025 11:43:32 +0200
Subject: [PATCH] Display Player


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java
index d0ab3564928fc9ae1a1fc043d0e841f11b5251d7..4d069b9a3ae608b31b0ee895c2b5129ac859d8ce 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java
@@ -477,6 +477,14 @@ public final class CraftEntityTypes {
         register(new EntityTypeData<>(EntityType.TNT_MINECART, ExplosiveMinecart.class, CraftMinecartTNT::new, createMinecart(net.minecraft.world.entity.EntityType.TNT_MINECART)));
         register(new EntityTypeData<>(EntityType.HOPPER_MINECART, HopperMinecart.class, CraftMinecartHopper::new, createMinecart(net.minecraft.world.entity.EntityType.HOPPER_MINECART)));
         register(new EntityTypeData<>(EntityType.SPAWNER_MINECART, SpawnerMinecart.class, CraftMinecartMobSpawner::new, createMinecart(net.minecraft.world.entity.EntityType.SPAWNER_MINECART)));
+        // kitpvp start - DisplayPlayer
+        register(new EntityTypeData<>(EntityType.DISPLAY_PLAYER, world.kitpvp.slime.api.entity.DisplayPlayer.class,
+            (server, displayPlayer) -> displayPlayer.getDefaultServerPlayer().getBukkitEntity(),
+            spawnData -> new world.kitpvp.slime.entity.DisplayPlayer(
+                spawnData.minecraftWorld(), spawnData.x(), spawnData.y(), spawnData.z()
+            ))
+        );
+        // kitpvp end - DisplayPlayer
 
         // None spawn able
         register(new EntityTypeData<>(EntityType.FISHING_BOBBER, FishHook.class, CraftFishHook::new, null)); // Cannot spawn a fish hook
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index cbbdc0acfb298602b22f2ec78485a45bb71912da..fa8dd1747045cb28dcd1dd013369788a7bca1a86 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -237,11 +237,21 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         this.firstPlayed = System.currentTimeMillis();
     }
 
+    // kitpvp start - Display player
     @Override
     public ServerPlayer getHandle() {
+        if(this.entity instanceof world.kitpvp.slime.entity.DisplayPlayer displayPlayer) {
+            return displayPlayer.getDefaultServerPlayer();
+        }
         return (ServerPlayer) this.entity;
     }
 
+    @Override
+    public boolean isBotEntity() {
+        return this.getHandle() instanceof world.kitpvp.slime.entity.DisplayServerPlayer;
+    }
+    // kitpvp end - Display player
+
     @Override
     public boolean equals(Object obj) {
         // Long-term, this should just use the super equals... for now, check the UUID
