From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: notstevy <notstevy@ultrabuildmc.de>
Date: Tue, 8 Apr 2025 20:49:40 +0200
Subject: [PATCH] PermissionBase Event API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 8fba3cbac5d2d39e7db3d3261c68317dbc936736..c6997597609affca32f034e71223ae0667d20504 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -89,7 +89,7 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
     private static final Logger LOGGER = LogUtils.getLogger();
     private CraftInventoryPlayer inventory;
     private final CraftInventory enderChest;
-    protected final PermissibleBase perm = new PermissibleBase(this);
+    protected final PermissibleBase perm;  // kitpvp - permissions api
     private boolean op;
     private GameMode mode;
 
@@ -98,6 +98,11 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         this.mode = server.getDefaultGameMode();
         this.inventory = new CraftInventoryPlayer(entity.getInventory());
         this.enderChest = new CraftInventory(entity.getEnderChestInventory());
+        // kitpvp - start: permission api
+        var event = new world.kitpvp.slime.api.event.PermissionsSetupEvent(this);
+        event.callEvent();
+        this.perm = event.permissibleBase();
+        // kitpvp - end
     }
 
     @Override
