From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: notstevy <notstevy@ultrabuildmc.de>
Date: Sun, 1 Sep 2024 18:01:49 +0200
Subject: [PATCH] Extend collides API - for kit duels custom spectator mode
 (players weren't able to place end crystals inside spectators)


diff --git a/src/main/java/net/minecraft/world/entity/EntitySelector.java b/src/main/java/net/minecraft/world/entity/EntitySelector.java
index dca7b99e97f21bf6cfae6ee69eeac95d0bcf6863..ebd4f174000fda395712ae22cd6b3621fb5dc683 100644
--- a/src/main/java/net/minecraft/world/entity/EntitySelector.java
+++ b/src/main/java/net/minecraft/world/entity/EntitySelector.java
@@ -25,6 +25,11 @@ public final class EntitySelector {
         return !(entity instanceof Player) || !entity.isSpectator() && !((Player) entity).isCreative();
     };
     public static final Predicate<Entity> NO_SPECTATORS = (entity) -> {
+        // kitpvp start - Extend collides API
+        if(entity instanceof LivingEntity livingEntity && !livingEntity.collides)
+            return false;
+        // kitpvp end - Extend collides API
+
         return !entity.isSpectator();
     };
     public static final Predicate<Entity> CAN_BE_COLLIDED_WITH = EntitySelector.NO_SPECTATORS.and(Entity::canBeCollidedWith);
